### PR TITLE
Make load snapshot streamable

### PR DESCRIPTION
**Pull request by pelmered**
This PR is making load snapshots streamable which is useful for loading larger database dumps. Currently the whole SQL dump is loaded into a variable and that causes out of memory errors. 

There are no tests, and the code could probably be improved. 
Some additional testing would also be great.

### CODE REVIEW

File: src/Commands/Load.php [hunk 1]
```diff
@@ -40,7 +40,11 @@ public function handle()
             return;
         }
 
-        $snapshot->load($this->option('connection'));
+        if ($this->option('stream')) {
+            $snapshot->loadStream($this->option('connection'));
+        } else {
```
Comment (freekmurze) [inline src/Commands/Load.php:45]: Avoid using else by using a ternary.

File: src/Snapshot.php [hunk 2]
```diff
@@ -61,6 +61,103 @@ public function load(string $connectionName = null)
...
+
+        if ($connectionName !== null) {
+            DB::setDefaultConnection($connectionName);
+        }
+
+        $this->dropAllCurrentTables();
+
+        if ($this->compressionExtension === 'gz') {
+            $dumpFilePath = $this->decompress();
+        } else {
```
Comment (freekmurze) [inline src/Snapshot.php:76]: Please avoid using an `else`

File: src/Snapshot.php [hunk 3]
```diff
@@ -61,6 +61,103 @@ public function load(string $connectionName = null)
...
+    {
+        if ($connectionName !== null) {
+            DB::setDefaultConnection($connectionName);
+        }
+
+        $tmpLine = '';
+        $lines = file($path);
+        $errors = [];
+
+        foreach ($lines as $line) {
```
Comment (freekmurze) [inline src/Snapshot.php:95]: Can we use a collection method here?

File: src/Snapshot.php [hunk 4]
```diff
@@ -61,6 +61,103 @@ public function load(string $connectionName = null)
...
+
+            // Skip it if line is a comment
```
Comment (freekmurze) [inline src/Snapshot.php:97]: Extract to a function so we can lose the comment

File: src/Snapshot.php [hunk 5]
```diff
@@ -61,6 +61,103 @@ public function load(string $connectionName = null)
...
+            if (substr($line, 0, 2) == '--' || trim($line) == '') {
```
Comment (freekmurze) [inline src/Snapshot.php:98]: Use strict checks `=== `

File: src/Snapshot.php [hunk 6]
```diff
@@ -61,6 +61,103 @@ public function load(string $connectionName = null)
...
+                continue;
+            }
+
+            // Add this line to the current segment
+            $tmpLine .= $line;
+
+            // If the line ends with a semicolon, it is the end of the query - run it
```
Comment (freekmurze) [inline src/Snapshot.php:105]: Avoid comments, but splitting this in to functions with a descriptive name

File: src/Snapshot.php [hunk 7]
```diff
@@ -61,6 +61,103 @@ public function load(string $connectionName = null)
...
+            if (substr(trim($line), -1, 1) == ';') {
+                try {
+                    DB::connection($connectionName)->unprepared($tmpLine);
+                } catch (\Exception $e) {
```
Comment (freekmurze) [inline src/Snapshot.php:109]: Import this so we can avoid that `\`

**Comment by pelmered** [comment]
Good feedback. It was mostly a proof of concept and I was aware that the code could be improved. It's great to hear that it's good functionality, because then I can spend some more time on improving the code and add tests.
I will try to fix most of this during the weekend.

File: src/Snapshot.php (hunk 3 above)
Comment (pelmered) [reply src/Snapshot.php:99]: How would I do this? Should I run `collect()` the `$lines` and then use the collection methods? I'm not sure that works in regard to memory. I can give it a try.

File: src/Snapshot.php (hunk 7 above)
Comment (pelmered) [reply src/Snapshot.php]: PHPStorm give me warnings about this without the `\` even if I import the exception.

**Comment by pelmered** [comment]
Left to do:
- The method `streamFileIntoDB` and `streamDecompress` in the `Snapshot` class needs to be rewritten. I think I have a pretty good idea on how to cleanup `streamDecompress`, but `streamFileIntoDB` is a bit more tricky.
- Add some unit tests.

**Comment by freekmurze** [comment]
Ping me when the above has been done. 👍

File: src/Snapshot.php (hunk 7 above)
Comment (bilfeldt) [reply src/Snapshot.php]: PHPStorm does not give warning for this if importing the proper exception as the top like so:

```php
use Exception;
...
try {
    DB::connection($connectionName)->unprepared($tmpLine);
} catch (Exception $e) {
...
}

File: src/Snapshot.php [hunk 8]
```diff
@@ -61,6 +61,103 @@ public function load(string $connectionName = null)
...
+            // Skip it if line is a comment
+            if (substr($line, 0, 2) == '--' || trim($line) == '') {
+                continue;
+            }
+
+            // Add this line to the current segment
+            $tmpLine .= $line;
+
+            // If the line ends with a semicolon, it is the end of the query - run it
+            if (substr(trim($line), -1, 1) == ';') {
```
Comment (bilfeldt) [inline src/Snapshot.php]: I suggest changing to `Str::endsWith(trim($line), ';')` for better readability

File: src/Snapshot.php (hunk 5 above)
Comment (bilfeldt) [inline src/Snapshot.php]: I suggest changing to `Str::startsWith($line, '--') || trim($line) == ''` for better readability

File: src/Snapshot.php (hunk 3 above)
Comment (bilfeldt) [reply src/Snapshot.php:99]: I suggest that we refactor this to a Lazy Collection method which is supported in Laravel 6+. This is taken from the [Lazy Collections documentation](https://laravel.com/docs/7.x/collections#creating-lazy-collections)

```php
use Illuminate\Support\LazyCollection;

LazyCollection::make(function () {
    $handle = fopen('log.txt', 'r');

    while (($line = fgets($handle)) !== false) {
        yield $line;
    }
});
```

probably a better way is streaming the content line for line exist using the FlySystem:
```php
$stream = $this->disk->readStream($path);
```
we can then refactor this to a function for streaming the content SQL line for line, meaning line ending with a semicolon.

**Comment by freekmurze** [comment]
Let's indeed use 

```
$this->disk->readStream($path);
```

for streaming.

**Comment by pelmered** [comment]
Thanks for the great feedback @bilfeldt! 
I did not know about LazyCollections, but that seems like the way to go.

**Comment by freekmurze** [comment]
There are still a few comments in my code review, that need to be addressed when you have got some time.

**Pull request closed without merging**